git-repair.1
.TH  
.SH NAME
git\-repair \- Fix a broken git repository
.PP
.SH SYNOPSIS
git\-repair [\-\-force]
.PP
.SH DESCRIPTION
This can fix a corrupt or broken git repository, which git fsck would
only complain has problems.
.PP
It does by deleting all corrupt objects, and retreiving all missing
objects that it can from the remotes of the repository.
.PP
If that is not sufficient to fully recover the repository, it can also
reset branches back to commits before the corruption happened, delete
branches that are no longer available due to the lost data, and remove any
missing files from the index. It will only do this if run with the
\fB\-\-force\fP option, since that rewrites history and throws out missing data.
Note that the \fB\-\-force\fP option never touches tags, even if they are no
longer usable due to missing data.
.PP
After running this command, you will probably want to run \fBgit fsck\fP to
verify it fixed the repository. Note that fsck may still complain about
objects referenced by the reflog, or the stash, if they were unable to be
recovered. This command does not try to clean up either the reflog or the
stash.
.PP
Since this command unpacks all packs in the repository, you may want to
run \fBgit gc\fP afterwards.
.SH TESTING
git-repair is able to test itself, by making a temporary copy
of the git reposiory it's run in, damaging it in random ways, and checking
that it can repair it so that git fsck reports no problems.
.PP
This is done using the --test parameter and associated --numtests and
--retry parameters.
.PP
Note that the testing will sometimes find repositories that can only be
repaired when --force is used; this is expected. On the other hand,
if it fails to recover a repository with --force, then the testing has
found a bug!
.PP
.SH AUTHOR
Joey Hess <joey@kitenet.net>
.PP
<http://git\-repair.branchable.com/>
.PP
.PP


debian/changelog
git-repair (1.20140227) unstable; urgency=medium

  * Optimise unpacking of pack files, and avoid repeated error
    messages about corrupt pack files.
  * Add swapping 2 files test case.

 -- Joey Hess <joeyh@debian.org>  Thu, 27 Feb 2014 11:56:27 -0400

git-repair (1.20140115) unstable; urgency=medium

  * Support old git versions from before git fsck --no-dangling was
    implemented.
  * Fix bug in packed refs file exploding code that caused a .gitrefs
    directory to be created instead of .git/refs
  * Check git version at run time.

 -- Joey Hess <joeyh@debian.org>  Wed, 15 Jan 2014 16:53:30 -0400

git-repair (1.20131213) unstable; urgency=low

  * Improve repair of index files in some situations.

 -- Joey Hess <joeyh@debian.org>  Fri, 13 Dec 2013 14:51:51 -0400

git-repair (1.20131203) unstable; urgency=low

  * Fix build deps. Closes: #731179

 -- Joey Hess <joeyh@debian.org>  Tue, 03 Dec 2013 15:02:21 -0400

git-repair (1.20131122) unstable; urgency=low

  * Added test mode, which can be used to randomly corrupt test 
    repositories, in reproducible ways, which allows easy
    corruption-driven-development.
  * Improve repair code in the case where the index file is corrupt,
    and this hides other problems.
  * Write a dummy .git/HEAD if the file is missing or corrupt, as
    git otherwise will not treat the repository as a git repo.
  * Improve fsck code to find badly corrupted objects that crash git fsck
    before it can complain about them.
  * Fixed crashes on bad file encodings.
  * Can now run 10000 tests (git-repair --test -n 10000 --force)
    with 0 failures.

 -- Joey Hess <joeyh@debian.org>  Fri, 22 Nov 2013 11:16:03 -0400

git-repair (1.20131118) unstable; urgency=low

  * First release

 -- Joey Hess <joeyh@debian.org>  Mon, 18 Nov 2013 13:38:12 -0400


git-repair.hs
{- git-repair program
 -
 - Copyright 2013 Joey Hess <joey@kitenet.net>
 -
 - Licensed under the GNU GPL version 3 or higher.
 -}

import Options.Applicative

import Common
import qualified Git.CurrentRepo
import qualified Git.Repair
import qualified Git.Config
import qualified Git.Construct
import qualified Git.Destroyer
import qualified Git.Fsck
import Utility.Tmp

data Settings = Settings
	{ forced :: Bool
	, testMode :: Bool
	, retryTestMode :: Bool
	, numTests :: Int
	}

parseSettings :: Parser Settings
parseSettings = Settings
	<$> switch
		( long "force"
		<> help "Force repair, even if data is lost"
		)
	<*> switch
		( long "test"
		<> help "Clone local repo, damage the clone, and test repair"
		)
	<*> switch
		( long "retry"
		<> help "Retry tests in git-repair-test.log"
		)
	<*> option
		( long "numtests"
		<> short 'n'
		<> metavar "N"
		<> help "Run N tests"
		<> value 1
		)

main :: IO ()
main = execParser opts >>= go
  where
  	opts = info (helper <*> parseSettings) desc
	desc = fullDesc
		<> header "git-repair - repair a damanged git repository" 
	go settings
		| retryTestMode settings = retryTest settings
		| testMode settings = test settings
		| otherwise = repair settings

repair :: Settings -> IO ()
repair settings = do
	g <- Git.Config.read =<< Git.CurrentRepo.get
	ifM (Git.Repair.successfulRepair <$> Git.Repair.runRepair Git.Repair.isTrackingBranch (forced settings) g)
		( exitSuccess
		, exitFailure
		)

test :: Settings -> IO ()
test settings = do
	forM_ [1 .. numTests settings] $ \n -> do
		putStrLn $ "** Test " ++ show n ++ "/" ++ show (numTests settings)
		damage <- Git.Destroyer.generateDamage
		logDamage damage
		runTest settings damage
	allOk

retryTest :: Settings -> IO ()
retryTest settings = do
	l <- map Prelude.read . lines <$> readFile logFile
	forM_ l $ \damage -> 
		runTest settings damage
	allOk

runTest :: Settings -> [Git.Destroyer.Damage] -> IO ()
runTest settings damage = withTmpDir "tmprepo" $ \tmpdir -> do
	let cloneloc = tmpdir </> "clone"
	cloned <- boolSystem "git"
		[ Param "clone"
		, Param "--no-hardlinks"
		, File "."
		, File cloneloc
		]
	unless cloned $
		error $ "failed to clone this repo"
	g <- Git.Config.read =<< Git.Construct.fromPath cloneloc
	Git.Destroyer.applyDamage damage g
	repairstatus <- catchMaybeIO $ Git.Repair.successfulRepair
		<$> Git.Repair.runRepair Git.Repair.isTrackingBranch (forced settings) g
	case repairstatus of
		Just True -> testResult repairstatus 
			. Just . not . Git.Fsck.foundBroken
			=<< Git.Fsck.findBroken False g
		_ -> testResult repairstatus Nothing

-- Pass test result and fsck result
testResult :: (Maybe Bool) -> (Maybe Bool) -> IO ()
testResult (Just True) (Just True) = putStrLn "** repair succeeded"
testResult (Just True) (Just False) = error "** repair succeeded, but final fsck failed"
testResult _ _ = error "** repair failed"

allOk :: IO ()
allOk = do
	putStrLn ""
	putStrLn "All tests ok!"

logDamage :: [Git.Destroyer.Damage] -> IO ()
logDamage damage = appendFile logFile $ show damage ++ "\n"

logFile :: FilePath
logFile = "git-repair-test.log"
